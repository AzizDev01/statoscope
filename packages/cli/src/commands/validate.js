const fs = require('fs');
const path = require('path');
const { parseChunked } = require('@discoveryjs/json-ext');
const { prepareWithJora } = require('@statoscope/webpack-model');
const { jora: joraHelpers } = require('@statoscope/helpers');

function makeQueryValidator(query) {
  function validate(type, message) {
    if (!(!type || type === 'error' || type === 'warn' || type === 'info')) {
      throw new Error(`Unknown message type [${type}]`);
    }

    if (!message) {
      throw new Error(`Message must be specified`);
    }
  }

  function callAPI(api, { type, filename, message }) {
    validate(type, message);

    if (typeof type === 'undefined' || type === 'error') {
      api.error(message, filename);
    } else if (type === 'warn') {
      api.warn(message, filename);
    } else if (type === 'info') {
      api.warn(message, filename);
    } else {
      console.log('Unknown message type:', type);
      api.error(message, filename);
    }
  }

  return (data, api) => {
    const result = data.query(query);

    for (const item of result) {
      if (!item.assert) {
        callAPI(api, { type: item.type, message: item.message, filename: item.filename });
      }
    }
  };
}

function handleValidator(validator) {
  const validatorPath = path.resolve(validator);
  const ext = path.extname(validatorPath);

  if (ext === '.js') {
    validator = require(validatorPath);
  } else {
    validator = fs.readFileSync(validatorPath, 'utf8');
  }

  if (typeof validator === 'string') {
    validator = makeQueryValidator(validator);
  }

  return validator;
}

module.exports = function (yargs) {
  return yargs.command(
    'validate [validator] [input]',
    `[BETA] Validate one or more JSON stats
Examples:
Single stats: validate path/to/validator.js path/to/stats.json
Multiple stats: generate path/to/validator.js --input path/to/stats-1.json path/to/stats-2.json`,
    (yargs) => {
      return yargs
        .positional('validator', {
          describe: 'path to validator script',
          alias: 'v',
          type: 'string',
        })
        .positional('input', {
          describe: 'path to a stats.json',
          alias: 'i',
          type: 'string',
        })
        .option('warn-as-error', {
          type: 'boolean',
          describe: 'Treat warnings as errors',
          alias: 'w',
        })
        .array('input')
        .demandOption(['validator', 'input']);
    },
    async (argv) => {
      const files = [];
      for (const file of argv.input) {
        const data = await parseChunked(fs.createReadStream(file));
        files.push({
          name: path.basename(file),
          data,
        });
      }
      const prepared = prepareWithJora(files, { helpers: joraHelpers() });
      const validator = handleValidator(argv.validator);
      const storage = {};
      let hasErrors = false;
      let errors = 0;
      let warnings = 0;
      let infos = 0;
      const api = {
        warn(message, filename = 'unknown') {
          if (argv['warn-as-error']) {
            hasErrors = true;
          }

          storage[filename] = storage[filename] || [];
          storage[filename].push({ type: 'warn', filename, message });
          warnings++;
        },
        error(message, filename = 'unknown') {
          hasErrors = true;
          storage[filename] = storage[filename] || [];
          storage[filename].push({ type: 'error', filename, message });
          errors++;
        },
        info(message, filename = 'unknown') {
          storage[filename] = storage[filename] || [];
          storage[filename].push({ type: 'info', filename, message });
          infos++;
        },
      };

      await validator({ files: prepared.files, query: prepared.query }, api);

      for (const [filename, items] of Object.entries(storage)) {
        console.log(filename);

        for (const item of items) {
          console.log(`  ${item.type}: ${item.message}`);
        }
      }

      console.log('');

      if (errors) {
        console.log(`Errors: ${errors}`);
      }

      if (warnings) {
        console.log(`Warnings: ${warnings}`);
      }
      if (infos) {
        console.log(`Infos: ${infos}`);
      }

      if (hasErrors) {
        // eslint-disable-next-line no-process-exit
        process.exit(1);
      }
    }
  );
};
