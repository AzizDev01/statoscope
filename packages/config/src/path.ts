import path from 'path';
import module from 'module';

export function normalizePath(source: string, rootDir: string): string {
  return source.replace('<rootDir>', rootDir);
}

export function makeRequireFromPath(pathname: string): NodeRequire {
  return module.createRequire(path.resolve(pathname, '_'));
}

export function resolveAliasPackage(
  prefixes: string[],
  name: string,
  fromDir: string
): string {
  const localRequire = makeRequireFromPath(fromDir);
  name = normalizePath(name, fromDir);

  if (name.startsWith('.') || path.isAbsolute(name)) {
    localRequire(name);
    return name;
  }

  let packageNamespace: string;
  let packageName: string;

  if (name.startsWith('@')) {
    [, packageNamespace = '', packageName = ''] = name.match(/(@.+?)[\\/](.+)/) ?? [];
  } else {
    packageNamespace = '';
    packageName = name;
  }

  const paths = [
    [name],
    ...prefixes.map((prefix) => [packageNamespace, `${prefix}-${packageName}`]),
  ].map((item) => path.join(...(item.filter(Boolean) as string[])));

  for (const item of paths) {
    try {
      localRequire(item);
      return item;
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  throw new Error(
    `Can't resolve ${name} with prefixes [${prefixes}].${
      !name.startsWith('.') ? 'Did you forget to install it?' : ''
    }`
  );
}
