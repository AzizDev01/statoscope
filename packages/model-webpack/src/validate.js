const { validate } = require('ajv');
const schema = require('./schema/stats.json');

export default function validateStats(stats) {
  try {
    validate(schema, stats);
    return { result: true };
  } catch (e) {
    return { result: false, message: e.message };
  }
}
