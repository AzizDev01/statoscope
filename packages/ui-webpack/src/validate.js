import { validate } from 'schema-utils';
import schema from '../schema/stats.json';

export default function validateStats(stats) {
  const configuration = { name: 'Stats' };

  try {
    validate(schema, stats, configuration);
    return { result: true };
  } catch (e) {
    return { result: false, message: e.message };
  }
}
