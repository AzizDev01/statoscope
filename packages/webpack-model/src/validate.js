import Ajv from 'ajv';
import schema from '../schema/stats.json';

export default function validateStats(stats) {
  const ajv = new Ajv();
  const validate = ajv.compile(schema);
  const valid = validate(stats);

  if (!valid) {
    return { result: false, errors: validate.errors };
  }

  return { result: true };
}
