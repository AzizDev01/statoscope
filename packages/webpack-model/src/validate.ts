import Ajv, { ErrorObject } from 'ajv';
import { Webpack } from '../webpack';
import schema from './schema/stats.json';

export type ValidationResult =
  | {
      result: false;
      errors: ErrorObject[];
    }
  | { result: true };

export default function validateStats(stats: Webpack.Compilation): ValidationResult {
  const ajv = new Ajv();
  const validate = ajv.compile(schema);
  const valid = validate(stats);

  if (!valid) {
    return { result: false, errors: validate.errors as ErrorObject[] };
  }

  return { result: true };
}
