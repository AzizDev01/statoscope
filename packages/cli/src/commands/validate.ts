import path from 'path';
import { Argv } from 'yargs';
import Validator from '@statoscope/stats-validator';
import ConsoleReporter from '@statoscope/stats-validator-reporter-console';
import { Config } from '@statoscope/stats-validator/dist/config';
import { ValidationResult } from '@statoscope/types/types/validation';
import legacyWebpackStatsValidator from './legacyWebpackValidator';

export default function (yargs: Argv): Argv {
  return yargs.command(
    'validate [validator] [input]',
    `[BETA] Validate one or more JSON stats
Examples:
Single stats: validate path/to/validator.js path/to/stats.json
Multiple stats: generate path/to/validator.js --input path/to/stats-1.json path/to/stats-2.json`,
    (yargs) => {
      return yargs
        .positional('validator', {
          describe: 'path to validator script',
          alias: 'v',
          type: 'string',
        })
        .positional('input', {
          describe: 'path to a stats.json',
          alias: 'i',
          type: 'string',
        })
        .option('config', {
          describe: 'path to statoscope config',
          alias: 'c',
          type: 'string',
        })
        .option('reference', {
          describe: 'path to a stats-file to diff with',
          alias: 'r',
          type: 'string',
        })
        .option('diff-with', {
          describe: 'path to a stats that will be used by diff-rules as "before"',
          type: 'string',
          alias: 'd',
        })
        .option('warn-as-error', {
          type: 'boolean',
          describe: 'Treat warnings as errors',
          alias: 'w',
        })
        .demandOption(['input']);
    },
    async (argv) => {
      let result: ValidationResult;

      if (argv['diff-with'] && argv.input.length > 1) {
        console.warn(
          'When "diff-with" is used, only the first file from "input" will be used'
        );
      }

      // todo statoscope 6: remove this or use as prepare + validator
      if (argv.validator) {
        result = await legacyWebpackStatsValidator(
          path.relative(process.cwd(), argv.validator),
          argv.input,
          argv.reference,
          {
            warnAsError: argv['warn-as-error'],
          }
        );
        const reporter = new ConsoleReporter();
        await reporter.run(result);
      } else {
        // todo resole nearest config
        const configPath = path.resolve(argv.config);
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const config = require(configPath) as Config;
        const validator = new Validator(config, path.dirname(configPath));

        result = await validator.validate(
          path.resolve(argv.input),
          argv.reference ? path.resolve(argv.reference) : void 0
        );

        for (const item of validator.reporters) {
          await item.run(result);
        }
      }
    }
  );
}
