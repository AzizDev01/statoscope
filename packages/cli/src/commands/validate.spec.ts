import os from 'os';
import path from 'path';
import yargs from 'yargs';
import validate from './validate';

const validatorFixturesJoraQuery = [
  '../../../../test/fixtures/cli/validate/ok.jora',
  '../../../../test/fixtures/cli/validate/fail.jora',
  '../../../../test/fixtures/cli/validate/rules.jora',
].map((filename) => ({
  name: path.basename(filename),
  path: path.resolve(__dirname, filename),
}));

const validatorFixturesJSJoraQuery = [
  '../../../../test/fixtures/cli/validate/jora-ok.js',
  '../../../../test/fixtures/cli/validate/jora-fail.js',
].map((filename) => ({
  name: path.basename(filename),
  path: path.resolve(__dirname, filename),
}));

const configFixturePath = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/validate/config.js'
);

const configReferenceFixturePath = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/validate/config-reference.js'
);

const inputFixturePath = path.resolve(
  __dirname,
  '../../../../test/bundles/simple/stats-dev.json'
);

const referenceFixturePath = path.resolve(
  __dirname,
  '../../../../test/bundles/v4/simple/stats-dev.json'
);

const rootPath = path.resolve(__dirname, '../../../../');
const outputDir = path.join(rootPath, 'test/temp', path.relative(rootPath, __filename));
// const consoleLog = console.log.bind(console);
const output: string[][] = [];
let consoleLogSpy: jest.SpyInstance<void, string[]> | null = null;

beforeEach(() => {
  consoleLogSpy = jest
    .spyOn(global.console, 'log')
    .mockImplementation((...args: string[]) => {
      output.push(
        args.map((o) =>
          String(o).replace(process.cwd(), '<pwd>').replace(os.tmpdir(), '<tmp>')
        )
      );
    });
});

afterEach(() => {
  consoleLogSpy?.mockRestore();
  output.length = 0;
});

function getOutput(): string[][] {
  return output;
}

describe('validator types', () => {
  test.each(validatorFixturesJoraQuery)('raw jora-query $name', async (item) => {
    let y = yargs(['validate', '--validator', item.path, '--input', inputFixturePath]);

    y = validate(y);
    y.fail((_, error) => {
      console.error(error);
    });

    await y.argv;

    expect(getOutput()).toMatchSnapshot();
  });

  test.each(validatorFixturesJSJoraQuery)('jora-query from js $name', async (item) => {
    let y = yargs(['validate', '--validator', item.path, '--input', inputFixturePath]);

    y = validate(y);
    y.fail((_, error) => {
      console.error(error);
    });

    await y.argv;

    expect(getOutput()).toMatchSnapshot();
  });

  test('with reference', async () => {
    let y = yargs([
      'validate',
      '--validator',
      validatorFixturesJoraQuery[2].path,
      '--input',
      inputFixturePath,
      '--reference',
      referenceFixturePath,
    ]);

    y = validate(y);
    y.fail((_, error) => {
      console.error(error);
    });

    await y.argv;

    expect(getOutput()).toMatchSnapshot();
  });
});

describe('with config', () => {
  describe('explicit', () => {
    test('single', async () => {
      let y = yargs([
        'validate',
        '--config',
        configFixturePath,
        '--input',
        inputFixturePath,
      ]);

      y = validate(y);
      y.fail((_, error) => {
        console.error(error);
      });

      await y.argv;

      expect(getOutput()).toMatchSnapshot();
    });

    test('reference', async () => {
      let y = yargs([
        'validate',
        '--config',
        configReferenceFixturePath,
        '--input',
        inputFixturePath,
        '--reference',
        referenceFixturePath,
      ]);

      y = validate(y);
      y.fail((_, error) => {
        console.error(error);
      });

      await y.argv;

      expect(getOutput()).toMatchSnapshot();
    });
  });
});
