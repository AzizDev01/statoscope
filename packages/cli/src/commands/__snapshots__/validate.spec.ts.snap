// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator types jora-query from js jora-fail.js 1`] = `
Array [
  Array [
    "unknown file",
  ],
  Array [
    "unknown compilation",
  ],
  Array [
    "error  Fail  <pwd>/test/fixtures/cli/validate/jora-fail.js",
  ],
  Array [
    "",
  ],
  Array [
    "1 problems (1 errors, 0 warnings)",
  ],
]
`;

exports[`validator types jora-query from js jora-ok.js 1`] = `Array []`;

exports[`validator types raw jora-query fail.jora 1`] = `
Array [
  Array [
    "unknown file",
  ],
  Array [
    "unknown compilation",
  ],
  Array [
    "error  Fail  <pwd>/test/fixtures/cli/validate/fail.jora",
  ],
  Array [
    "",
  ],
  Array [
    "1 problems (1 errors, 0 warnings)",
  ],
]
`;

exports[`validator types raw jora-query ok.jora 1`] = `Array []`;

exports[`validator types raw jora-query rules.jora 1`] = `
Array [
  Array [
    "unknown file",
  ],
  Array [
    "unknown compilation",
  ],
  Array [
    "info  Auto info message  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "info  Info message without assert  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "info  Info with assert  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "warn  Warn without assert  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "warn  Warn with assert  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "",
  ],
  Array [
    "input.json",
  ],
  Array [
    "unknown compilation",
  ],
  Array [
    "info  Auto info for file  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "info  Info for file  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "warn  Warn for file  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "error  Error for file  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "da95abcf6bbf9b157ff6",
  ],
  Array [
    "error  Something wrong  <pwd>/test/fixtures/cli/validate/rules.jora",
  ],
  Array [
    "",
  ],
  Array [
    "5 problems (2 errors, 3 warnings)",
  ],
]
`;
